sway-core/src/type_system/type_argument.rs

```diff
@@ -56,6 +58,8 @@ impl From<&TypeParameter> for TypeArgument {
             type_id: type_param.type_id,
             initial_type_id: type_param.initial_type_id,
             span: type_param.name_ident.span(),
+            name_span: None,
+            type_arguments: vec![],
```

Comment:
Can you elaborate on why you have added this field? Or give a code example that would trigger this condition?

Reply:
Yeah this requires some explanation.

So let's say we're compiling `fun::<Option<Result<u32, u32>>>();` We want individual tokens that point to definitions for both `Option` and `Result` but the type system as it stands is only interested in figuring out what the type of `Option<Result<u32, u32>>` is. There is only a single instance of TypeArgument.

What we need is a way to store the entire tree so we can evaluate the type of every node (and store the associated `name_span`s). This is what this field allows: a `TypeArgument` potentially stores children that will have to be resolved as well.

The need for recursive evaluation of the arguments elsewhere might make more sense in light of that.

Reply:
Adding a `type_arguments` field on `TypeArgument` will create duplicated information---the type information in `type_id` contains the same types that would be `type_arguments`, granted without the `Span` information as you mentioned.

Greater than this problem though---is it accurate to say that the LSP is unable to find tokens for most types, in general?

Reply:
>Greater than this problem though---is it accurate to say that the LSP is unable to find tokens for most types, in general?

I don't believe it's accurate, there are missing tokens in some edge cases but it does produce the correct info in general.

>Adding a type_arguments field on TypeArgument will create duplicated information---the type information in type_id contains the same types that would be type_arguments, granted without the Span information as you mentioned.

I think I might be able to get away with just storing the spans in the `TypeArgument` and reconstruct the tree in the LSP, which should simplify this PR a lot if it works.